litestar/serialization.py

```diff
@@ -125,6 +129,38 @@ def default_serializer(value: Any, type_encoders: Mapping[Any, Callable[[Any], A
     raise TypeError(f"Unsupported type: {type(value)!r}")
 
 
+PydanticUUIDType = TypeVar("PydanticUUIDType", bound="UUID1")
+
+
+def _dec_pydantic_uuid(type_: type[PydanticUUIDType], val: Any) -> PydanticUUIDType:
+    if isinstance(val, str):
+        val = type_(val)
+    elif isinstance(val, (bytes, bytearray)):
+        try:
+            val = type_(val.decode())
+        except ValueError:
+            # 16 bytes in big-endian order as the bytes argument fail
+            # the above check
+            val = type_(bytes=val)
+    elif isinstance(val, UUID):
+        val = type_(str(val))
+
+    if not isinstance(val, type_):
+        raise ValidationError(f"Invalid UUID: {val!r}")
+
+    if type_._required_version != val.version:  # type:ignore[attr-defined]
+        raise ValidationError(f"Invalid UUID version: {val!r}")
+
+    return val
+
+
+def _dec_pydantic(type_: type[BaseModel], value: Any) -> BaseModel:
+    try:
+        return type_.parse_obj(value)
+    except PydanticValidationError as e:
+        raise ValidationError from e
```

Comment:
In [the docs](https://github.com/jcrist/msgspec/pull/445) it says we should raise `TypeError` in dec hooks, but that seemed to be raised directly and wouldn't be caught  by the error handling in the `parse_values_from_connection_kwargs()` method.

Reply:
> In https://github.com/jcrist/msgspec/pull/445 it says we should raise TypeError in dec hooks, but that seemed to be raised directly

*( the link you posted above is to a PR, I assume you meant to link something from [here](https://jcristharif.com/msgspec/extending.html#mapping-to-from-native-types))*

The docs don't actually take a stance on what you should raise on encode/decode error, the only error documented there is for raising on types that the `enc_hook`/`dec_hook` don't handle. I'm going to update the docs to recommend raising `NotImplementedError` in those cases instead of `TypeError`.

For fixing your issue here, I'll push a fix upstream to `msgspec` to wrap any `ValueError`/`TypeError`/`ValidationError` raised in a `msgspec.ValidationError` with the standard additional context. Since `PydanticValidationError` subclasses from `ValueError` this should be handled transparently.

I've opened https://github.com/jcrist/msgspec/issues/456 for tracking this.

Reply:
Thanks @jcrist !

Reply:
Thanks @jcrist.

The `TypeError` is referenced in the api docs for [`convert.dec_hook`](https://jcristharif.com/msgspec/api.html#msgspec.convert) (which is where the link was supposed to reference):

> dec_hook (callable, optional) – ... This hook should transform obj into type type, or raise a TypeError if unsupported.